### PR TITLE
Added torch.device and CWrapper in generic

### PR DESCRIPTION
**Pull request by madhavajay**
## Description
- Allows inheriting from types which cannot be inherited from
- Added upcast to SyPrimitives

## Updates
- Fixed issues with CWrapper and id property
- Fixed test failing due to new methods on PyPrimitive
- Fixed PyPrimitive return types in allowlist
- Fixed issue with print / repr(torch.nn.Parameter) throwing exception
- Fixed main return_type_name being squashed by methods return type
- Note: __id and __getattribute__ causes a Class property lookup
  which then fails. The solution is not to use __id but _id

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
Locally, WIP failing test

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/syft/core/node/common/action/function_or_constructor_action.py [hunk 1]
```diff
@@ -104,7 +104,24 @@ def execute_action(self, node: AbstractNode, verify_key: VerifyKey) -> None:
             )
             resolved_kwargs[arg_name] = r_arg.data
 
-        result = method(*resolved_args, **resolved_kwargs)
+        # when we call original constructors like torch.device we will get errors
```
Comment (iamtrask) [inline src/syft/core/node/common/action/function_or_constructor_action.py:107]: looks great!

also - didn't know this was called "upcast"ing - im a fan

File: src/syft/lib/generic.py [hunk 2]
```diff
@@ -16,6 +16,19 @@
 from .util import replace_classes_in_module
 
 
+class CWrapper:
```
Comment (iamtrask) [inline src/syft/lib/generic.py:19]: need more documentation on this class - please fully explain how it works

File: src/syft/lib/python/bool.py [hunk 3]
```diff
@@ -49,6 +49,10 @@ def id(self) -> UID:
         """
         return self._id
 
+    @syft_decorator(typechecking=True, prohibit_args=True)
```
Comment (iamtrask) [inline src/syft/lib/python/bool.py:52]: really go crazy on documentation for this one - it's a complex/abstract feature

Comment (iamtrask) [reply src/syft/lib/python/bool.py:52]: (same for the other ones - can probably be mostly copy psate)

**Review by iamtrask — CHANGES_REQUESTED** [review]
most docs

File: src/syft/core/node/common/action/function_or_constructor_action.py (hunk 1 above)
Comment (madhavajay) [reply src/syft/core/node/common/action/function_or_constructor_action.py:107]: Yeah I had to google it to double check... I was like... its not down casting... so whats the opposite?
There is probably a better mathy name for it somewhere like something to do with "contravariance" or something.

File: src/syft/lib/generic.py (hunk 2 above)
Comment (madhavajay) [reply src/syft/lib/generic.py:19]: Yup I will document next, still getting it to work unfortunately, hit some weird issues with property / method pointer return types and getattr on __id on the CWrapper.

**Comment by madhavajay** [comment]
The generic CWrapper / ShadowWrapper approach is no longer being used so this PR can be closed for now.
